File: Deepspace/lib/Hangar.rb
#encoding:utf-8

require_relative "Weapon.rb"
require_relative "ShieldBooster"
require_relative "enums"
require_relative "HangarToUI"

module Deepspace

class Hangar

	# Initializers
	#=======================================================================

	# Description:
	# 	Initializer of the class
	# Parameters:
	# 	capacity: Integer, max number of shieldBoosters and weapons
	def initialize(capacity)
		@maxElements = capacity
		@shieldBoosters = []
		@weapons = []
	end

	# Description:
	# 	Copy initializer
	# Param:
	# 	h: Hangar, the objet to be copied
	# Returns:
	# 	Hangar, a copy of the hangar parameter given
	def self.newCopy(h)
		copy = new(h.maxElements)

		for shieldBooster in h.shieldBoosters
			copy.addShieldBooster(shieldBooster)
		end

		for weapon in h.weapons
			copy.addWeapon(weapon)
		end

		return copy
	end

	# Getters
	#=======================================================================

	attr_reader :shieldBoosters, :weapons, :maxElements

	# Description:
	# 	Checks if there's space left for more elements at the hangar
	# Returns:
	# 		Boolean:	true, if there's space left for at least one more element
	# 					false, if the hangar is full
	def spaceAvailable
		return @maxElements > @weapons.length + @shieldBoosters.length
	end

	# Description:
	# 	Gets the UI representation of the object
	# Returns
	# 	HangarToUI: the associated UI representation 
	def getUIVersion
		return HangarToUI.new(self)
	end

	# Description:
	# 	Returns a string representation of the object
	# Returns:
	# 	String: a representation of the object
	def to_s
		return "[Hangar]-> Max. elements: #{@maxElements}, Weapons: [#{@weapons.join(", ")}],  Shields: [#{@shieldBoosters.join(", ")}]"
	end

	# Setters
	#=======================================================================

	# Description:
	# 	Adds a new weapon to the Hangar
	# Parameters:
	# 	w: Weapon, the weapon to be added
	# Returns:
	# 	Boolean: true, if the operation runs successfully
	# 			 false, if something fails (no room for another weapon)
	def addWeapon(w)
		if spaceAvailable
			@weapons << w
			return true
		else
			return false
		end
	end

	# Description:
	# 	Removes a weapon from the Hangar
	# Parameters:
	# 	w: Integer
	# Returns:
	# 	Weapon:	nil, if position is invalid
	# 			the weapon deleted, if position is valid
	def removeWeapon(w)
		if w >= @weapons.length || w < 0
			puts "WARNING! Invalid position to specify a weapon, on Hangar.removeWeapon(w: Integer)"
			return nil
		else
			return @weapons.delete_at(w)
		end
	end

	# Description:
	# 	Adds a new shieldBooster to the Hangar
	# Parameters:
	# 		s: ShielBooster, the shield to be added
	# Returns:
	# 		Boolean:	true, if the operation runs successfully
	# 					false, if something fails (no room for another shield)
	def addShieldBooster(s)
		if spaceAvailable
			@shieldBoosters << s
			return true
		else
			return false
		end
	end

	# Description:
	# 	Removes a shield from the Hangar
	# Parameters:
	# 	s: Integer, position of the shield to remove
	# Returns:
	# 	ShielBooster,	Nil, if position is not valid
	# 					the shieldBooster which has been deleted
	def removeShieldBooster(s)
		if s >= @shieldBoosters.length || s < 0
			puts "WARNING! Invalid position to specify a shield, on Hangar.removeShieldBooster(s: Integer)"
			return nil
		else
			return  @shieldBoosters.delete_at(s)
		end
	end

end # class Hangar

end # module Deepspace

File: deepspace/src/deepspace/Hangar.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.ArrayList;

/**
 *
 * @author Miguel Ángel Fernández Gutiérrez, Sergio Quijano Rey
 */
class Hangar {
    /**
     * Maximum number of shields and weapons --combined-- the hangar can have
     */
    private int maxElements;
    
    /**
     * Array of shield boosters the hangar has
     */
    private ArrayList<ShieldBooster> shieldBoosters;
    
    /**
     * Array of weapons the hangar has
     */
    private ArrayList<Weapon> weapons;
    
    // -------------------------------------------------------------------------
    // Constructors
    // -------------------------------------------------------------------------
    
    /**
     * Class initializer
     * @param capacity maximum number of shields and weapons the hangar can have
     */
    Hangar(int capacity) {
        maxElements = capacity;
        shieldBoosters = new ArrayList<>();
        weapons = new ArrayList<>();
    }
    
    /**
     * Copy constructor
     * @param h instance which is going to be copied
     */
    Hangar(Hangar h) {
        maxElements = h.maxElements;
        
        // Call to getter creates copy
        shieldBoosters = h.shieldBoosters;
        weapons = h.weapons;
    }
    
    // -------------------------------------------------------------------------
    // Getters
    // -------------------------------------------------------------------------
    
    /**
     * Getter for maxElements
     * @return maxElements
     */
    public int getMaxElements() {
        return maxElements;
    }
    
    /**
     * Getter for shieldBoosters
     * @return shieldBoosters
     */
    public ArrayList<ShieldBooster> getShieldBoosters() {
        // Reference protection
        /*ArrayList<ShieldBooster> shieldBoosters_copy = new ArrayList<>();
        for ( ShieldBooster sh : shieldBoosters )
            shieldBoosters_copy.add(sh);
        
        return shieldBoosters_copy;*/
        
        return shieldBoosters;
    }
    
    /**
     * Getter for weapons
     * @return weapons
     */
    public ArrayList<Weapon> getWeapons() {
        // Reference protection
        /*ArrayList<Weapon> weapons_copy = new ArrayList<>();
        for ( Weapon weap : weapons )
            weapons_copy.add(weap);
        
        return weapons_copy;*/
        
        return weapons;
    }
    
    /**
     * Checks if there's space left for more elements at the hangar
     * @return true, if there's space left for at least one more element;
     *         false, if the hangar is full
     */
    private boolean spaceAvailable() {
        return maxElements > shieldBoosters.size() + weapons.size();
    }
    
    // -------------------------------------------------------------------------
    // Setters
    // -------------------------------------------------------------------------
    
    /**
     * Adds a new weapon to the hangar
     * @param w the weapon to be added
     * @return true, if the operation runs successfully;
     *         false, if something fails (no room for another weapon)
     */
    public boolean addWeapon(Weapon w) {
        if ( spaceAvailable() )
            return weapons.add(w);
        else
            return false;
    }
    
    /**
     * Removes a weapon from the hangar
     * @param w the position in which the weapon that wants to be removed is located
     * @return null if position is invalid or removal operation is unsuccessful;
     *         the weapon, if removal is successful
     */
    public Weapon removeWeapon(int w) {
        if ( w >= weapons.size() )
            return null;
        else
            return weapons.remove(w);
    }
    
    /**
     * Adds a new shield booster to the hangar
     * @param w the shield booster to be added
     * @return true, if the operation runs successfully;
     *         false, if something fails (no room for another booster)
     */
    public boolean addShieldBooster(ShieldBooster s) {
        if ( spaceAvailable() )
            return shieldBoosters.add(s);
        else
            return false;
    }
    
    /**
     * Removes a shield booster from the hangar
     * @param w the position in which the booster that wants to be removed is located
     * @return null if position is invalid or removal operation is unsuccessful;
     *         the shield booster, if removal is successful
     */
    public ShieldBooster removeShieldBooster(int s) {
        ShieldBooster removed;
        if ( s >= shieldBoosters.size() )
            removed = null;
        else
            removed = shieldBoosters.remove(s);
        
        return removed;
    }
    
    // -------------------------------------------------------------------------
    // String representation, UI version
    // -------------------------------------------------------------------------
    
    /**
     * String representation of the object
     * @return string representation
     */
    public String toString() {
        String message = "[Hangar] -> Max. elements: " + maxElements
                + ", Shields: " + shieldBoosters.toString()
                + ", Weapons: " + weapons.toString();
        return message;
    }
    
    /**
     * To UI
     */
    HangarToUI getUIversion() {
        return new HangarToUI(this);
    }
}
